include/cufinufft/impl.h

```diff
@@ -76,205 +83,252 @@ int cufinufft_makeplan_impl(int type, int dim, int *nmodes, int iflag, int ntran
+                  d_plan->opts.gpu_obinsizex);
     if (dim > 1)
-      onedim_fseries_kernel_precomp(nf2, f + MAX_NQUAD, a + MAX_NQUAD, d_plan->spopts);
+      set_nf_type12(d_plan->mt, d_plan->opts, d_plan->spopts, &nf2,
+                    d_plan->opts.gpu_obinsizey);
     if (dim > 2)
-      onedim_fseries_kernel_precomp(nf3, f + 2 * MAX_NQUAD, a + 2 * MAX_NQUAD,
-                                    d_plan->spopts);
+      set_nf_type12(d_plan->mu, d_plan->opts, d_plan->spopts, &nf3,
+                    d_plan->opts.gpu_obinsizez);
+
+    // dynamically request the maximum amount of shared memory available
+    // for the spreader
+
+    /* Automatically set GPU method. */
+    if (d_plan->opts.gpu_method == 0) {
+      /* For type 1, we default to method 2 (SM) since this is generally faster
+       * if there is enough shared memory available. Otherwise, we default to GM.
+       *
+       * For type 2, we always default to method 1 (GM).
+       */
+      if (type == 2) {
+        d_plan->opts.gpu_method = 1;
+      } else {
+        // query the device for the amount of shared memory available
+        int shared_mem_per_block{};
+        cudaDeviceGetAttribute(&shared_mem_per_block,
+                               cudaDevAttrMaxSharedMemoryPerBlockOptin, device_id);
+        // compute the amount of shared memory required for the method
+        const auto shared_mem_required = shared_memory_required<T>(
+            dim, d_plan->spopts.nspread, d_plan->opts.gpu_binsizex,
+            d_plan->opts.gpu_binsizey, d_plan->opts.gpu_binsizez);
+        if ((shared_mem_required > shared_mem_per_block)) {
+          d_plan->opts.gpu_method = 1;
+        } else {
+          d_plan->opts.gpu_method = 2;
+        }
+      }
+    }
 
-    if ((ier = checkCudaErrors(
-             cudaMallocWrapper(&d_a, dim * MAX_NQUAD * sizeof(cuDoubleComplex), stream,
-                               d_plan->supports_pools))))
-      goto finalize;
-    if ((ier = checkCudaErrors(cudaMallocWrapper(&d_f, dim * MAX_NQUAD * sizeof(T),
-                                                 stream, d_plan->supports_pools))))
+    if ((ier = cudaGetLastError())) {
       goto finalize;
-    if ((ier = checkCudaErrors(
-             cudaMemcpyAsync(d_a, a, dim * MAX_NQUAD * sizeof(cuDoubleComplex),
-                             cudaMemcpyHostToDevice, stream))))
-      goto finalize;
-    if ((ier = checkCudaErrors(cudaMemcpyAsync(d_f, f, dim * MAX_NQUAD * sizeof(T),
-                                               cudaMemcpyHostToDevice, stream))))
+    }
+
+    d_plan->nf1 = nf1;
+    d_plan->nf2 = nf2;
+    d_plan->nf3 = nf3;
+    d_plan->nf  = nf1 * nf2 * nf3;
+    if (d_plan->opts.debug) {
+      printf("[cufinufft] (nf1,nf2,nf3) = (%d, %d, %d)\n", d_plan->nf1, d_plan->nf2,
+             d_plan->nf3);
+    }
+
+    using namespace cufinufft::memtransfer;
+    switch (d_plan->dim) {
+    case 1: {
+      if ((ier = allocgpumem1d_plan<T>(d_plan))) goto finalize;
```

Comment:
`goto` were already used in the code. I left it because it is a good way to do error management. `goto` can only be use for this particular use case and they are quite clear.